File: training/spinemetrics_training/preprocess.py
"""Preprocesado de imagen. SPEC.md §8 Etapa 2: "Normalización de
intensidad, CLAHE para realzar bordes corticales, detección de la ROI de
la columna, remuestreo a la resolución de entrada del modelo. Conservar la
transformación afín: todas las mediciones se calculan en el espacio de la
imagen original, nunca en el remuestreado."

`detect_spine_roi` es un heurístico de referencia (varianza de columna),
**no** el detector de ROI final del pipeline — ese exigiría su propio
modelo entrenado, fuera del alcance de este andamiaje (el usuario pidió
sólo preparar el entrenamiento, no la inferencia en `src/pipeline/`).
Está aquí para que el resto del preprocesado (recorte + remuestreo con
transformación afín) se pueda ejercitar y probar de extremo a extremo.
"""
from __future__ import annotations

from dataclasses import dataclass

import cv2
import numpy as np

from spinemetrics_training.schema import Point

Roi = tuple[int, int, int, int]  # (x0, y0, x1, y1), semiabierto: [x0,x1) × [y0,y1)


@dataclass(frozen=True)
class AffineTransform:
    """Mapea puntos del espacio ORIGINAL al espacio REMUESTREADO:
    `resampled = original * scale + offset` (por eje, sin rotación — el
    pipeline de SpineMetrics nunca rota la imagen, sólo recorta y
    escala)."""

    scale_x: float
    scale_y: float
    offset_x: float = 0.0
    offset_y: float = 0.0

    @staticmethod
    def identity() -> "AffineTransform":
        return AffineTransform(1.0, 1.0, 0.0, 0.0)

    def apply(self, point: Point) -> Point:
        """Original → remuestreado."""
        return Point(point.x * self.scale_x + self.offset_x, point.y * self.scale_y + self.offset_y)

    def apply_inverse(self, point: Point) -> Point:
        """Remuestreado → original. SPEC.md §13.7: "landmarks detectados en
        el espacio remuestreado y devueltos al original conservan la
        posición dentro de 0.5 px."""
        return Point((point.x - self.offset_x) / self.scale_x, (point.y - self.offset_y) / self.scale_y)

    def compose(self, other: "AffineTransform") -> "AffineTransform":
        """Transformación equivalente a aplicar primero `self` y luego
        `other`: `other.apply(self.apply(point))` para cualquier `point`."""
        return AffineTransform(
            scale_x=self.scale_x * other.scale_x,
            scale_y=self.scale_y * other.scale_y,
            offset_x=self.offset_x * other.scale_x + other.offset_x,
            offset_y=self.offset_y * other.scale_y + other.offset_y,
        )


def normalize_intensity(image: np.ndarray, low_percentile: float = 1.0, high_percentile: float = 99.0) -> np.ndarray:
    """Normaliza a `float32` en `[0, 1]` recortando por percentiles (robusto
    a valores atípicos de saturación/aire, frecuentes en radiografía)."""
    image = image.astype(np.float32)
    low, high = (float(v) for v in np.percentile(image, [low_percentile, high_percentile]))
    if high <= low:
        return np.zeros_like(image, dtype=np.float32)
    normalized = (image - low) / (high - low)
    return np.clip(normalized, 0.0, 1.0).astype(np.float32)


def apply_clahe(image_uint8: np.ndarray, clip_limit: float = 2.0, tile_grid_size: tuple[int, int] = (8, 8)) -> np.ndarray:
    """CLAHE (contrast-limited adaptive histogram equalization) para
    realzar bordes corticales (SPEC.md §8 Etapa 2). Espera `uint8`."""
    clahe = cv2.createCLAHE(clipLimit=clip_limit, tileGridSize=tile_grid_size)
    return clahe.apply(image_uint8)


def detect_spine_roi(image: np.ndarray, min_width_fraction: float = 0.3, margin_fraction: float = 0.15) -> Roi:
    """Heurístico de referencia: conserva la banda de columnas con mayor
    varianza de intensidad (donde se concentra el contraste cortical de la
    columna en una PA/lateral centrada), con un margen. Conserva siempre
    toda la altura — recortar craneal/caudal exige saber dónde empieza y
    termina la columna, que sí depende de un modelo real."""
    height, width = image.shape[:2]
    column_variance = image.astype(np.float32).var(axis=0)
    threshold = column_variance.max() * 0.2
    active_columns = np.where(column_variance >= threshold)[0]
    if len(active_columns) == 0:
        return (0, 0, width, height)

    x0, x1 = int(active_columns[0]), int(active_columns[-1]) + 1
    margin = int((x1 - x0) * margin_fraction)
    x0 = max(0, x0 - margin)
    x1 = min(width, x1 + margin)

    min_width = int(width * min_width_fraction)
    if x1 - x0 < min_width:
        center = (x0 + x1) // 2
        x0 = max(0, center - min_width // 2)
        x1 = min(width, x0 + min_width)

    return (x0, 0, x1, height)


def resample_with_affine(image: np.ndarray, roi: Roi, target_size: tuple[int, int]) -> tuple[np.ndarray, AffineTransform]:
    """Recorta a `roi` y remuestrea a `target_size = (ancho, alto)`,
    devolviendo también la `AffineTransform` que mapea el espacio original
    al remuestreado — SPEC.md §8 Etapa 2, "conservar la transformación
    afín"."""
    x0, y0, x1, y1 = roi
    cropped = image[y0:y1, x0:x1]
    cropped_height, cropped_width = cropped.shape[:2]
    if cropped_height == 0 or cropped_width == 0:
        raise ValueError(f"ROI vacía: {roi!r} sobre una imagen de forma {image.shape!r}.")

    target_width, target_height = target_size
    scale_x = target_width / cropped_width
    scale_y = target_height / cropped_height
    resized = cv2.resize(cropped, (target_width, target_height), interpolation=cv2.INTER_LINEAR)

    crop_transform = AffineTransform(1.0, 1.0, offset_x=-x0, offset_y=-y0)
    resize_transform = AffineTransform(scale_x, scale_y)
    return resized, crop_transform.compose(resize_transform)


def resample_mask_with_roi(mask_channels: np.ndarray, roi: Roi, target_size: tuple[int, int]) -> np.ndarray:
    """Aplica el mismo recorte + remuestreo que `resample_with_affine` a una
    máscara multicanal `(C, H, W)`, para que quede alineada píxel a píxel
    con la imagen remuestreada correspondiente. Usa interpolación por
    vecino más cercano (no lineal): una máscara binaria no debe adquirir
    valores fraccionarios por el remuestreo."""
    x0, y0, x1, y1 = roi
    num_channels = mask_channels.shape[0]
    target_width, target_height = target_size
    out = np.zeros((num_channels, target_height, target_width), dtype=np.float32)
    for channel in range(num_channels):
        cropped = mask_channels[channel, y0:y1, x0:x1]
        out[channel] = cv2.resize(cropped, (target_width, target_height), interpolation=cv2.INTER_NEAREST)
    return out

File: src/pipeline/preprocess.ts
/**
 * Preprocesado de imagen. SPEC.md §8 Etapa 2: "Normalización de intensidad,
 * CLAHE para realzar bordes corticales, detección de la ROI de la columna,
 * remuestreo a la resolución de entrada del modelo. Conservar la
 * transformación afín."
 *
 * Espejo en TypeScript de `training/spinemetrics_training/preprocess.py`
 * (mismo algoritmo de detección de ROI y de remuestreo con transformación
 * afín) para que el comportamiento de entrenamiento e inferencia se pueda
 * auditar en paralelo. Opera sobre `GrayscaleImage` puro (sin `ImageBitmap`
 * ni canvas) para poder probarse en Node — ver `pipeline/types.ts`.
 */
import type { AffineTransform2D, GrayscaleImage, SpineRoi } from './types';
import { composeAffine } from './types';

/** Normaliza a `[0, 1]` recortando por percentiles (robusto a saturación/
 * aire), igual que `preprocess.normalize_intensity` en Python. */
export function normalizeIntensity(image: GrayscaleImage, lowPercentile = 1, highPercentile = 99): GrayscaleImage {
  const sorted = Float32Array.from(image.data).sort();
  const low = percentile(sorted, lowPercentile);
  const high = percentile(sorted, highPercentile);

  const out = new Float32Array(image.data.length);
  if (high <= low) {
    return { width: image.width, height: image.height, data: out };
  }
  const range = high - low;
  for (let i = 0; i < image.data.length; i++) {
    const v = (image.data[i]! - low) / range;
    out[i] = v < 0 ? 0 : v > 1 ? 1 : v;
  }
  return { width: image.width, height: image.height, data: out };
}

/** Percentil sobre un array YA ordenado ascendentemente, por interpolación
 * lineal entre los dos índices más cercanos (método "linear" de numpy, el
 * mismo que usa `np.percentile` en `training/preprocess.py`). */
function percentile(sortedData: Float32Array, p: number): number {
  if (sortedData.length === 0) return 0;
  if (sortedData.length === 1) return sortedData[0]!;
  const rank = (p / 100) * (sortedData.length - 1);
  const lowerIndex = Math.floor(rank);
  const upperIndex = Math.ceil(rank);
  if (lowerIndex === upperIndex) return sortedData[lowerIndex]!;
  const t = rank - lowerIndex;
  return sortedData[lowerIndex]! * (1 - t) + sortedData[upperIndex]! * t;
}

/**
 * CLAHE (contrast-limited adaptive histogram equalization) implementado
 * directamente sobre `GrayscaleImage` (sin OpenCV.js, para no añadir una
 * dependencia nativa pesada al bundle del navegador). Espera una imagen ya
 * normalizada a `[0, 1]` (ver `normalizeIntensity`); devuelve otra en
 * `[0, 1]`. Algoritmo estándar: histograma por mosaico con recorte y
 * redistribución del exceso, interpolación bilineal entre los cuatro
 * mosaicos más cercanos a cada píxel.
 */
export function applyClahe(image: GrayscaleImage, tilesX = 8, tilesY = 8, clipLimit = 2.0): GrayscaleImage {
  const { width, height, data } = image;
  const bins = 256;
  const tileWidth = Math.max(1, Math.ceil(width / tilesX));
  const tileHeight = Math.max(1, Math.ceil(height / tilesY));
  const numTilesX = Math.ceil(width / tileWidth);
  const numTilesY = Math.ceil(height / tileHeight);

  // Un lookup table (CDF recortada, 256 valores en [0,1]) por mosaico.
  const luts: Float32Array[][] = [];
  for (let ty = 0; ty < numTilesY; ty++) {
    const row: Float32Array[] = [];
    for (let tx = 0; tx < numTilesX; tx++) {
      row.push(buildTileLut(data, width, tx * tileWidth, ty * tileHeight, tileWidth, tileHeight, width, height, bins, clipLimit));
    }
    luts.push(row);
  }

  const out = new Float32Array(data.length);
  for (let y = 0; y < height; y++) {
    // Centro del mosaico más cercano a cada lado, en coordenadas de mosaico continuas.
    const ty = (y - tileHeight / 2) / tileHeight;
    const ty0 = clampInt(Math.floor(ty), 0, numTilesY - 1);
    const ty1 = clampInt(ty0 + 1, 0, numTilesY - 1);
    const fy = clamp01(ty - ty0);

    for (let x = 0; x < width; x++) {
      const tx = (x - tileWidth / 2) / tileWidth;
      const tx0 = clampInt(Math.floor(tx), 0, numTilesX - 1);
      const tx1 = clampInt(tx0 + 1, 0, numTilesX - 1);
      const fx = clamp01(tx - tx0);

      const bin = clampInt(Math.round(data[y * width + x]! * (bins - 1)), 0, bins - 1);
      const v00 = luts[ty0]![tx0]![bin]!;
      const v01 = luts[ty0]![tx1]![bin]!;
      const v10 = luts[ty1]![tx0]![bin]!;
      const v11 = luts[ty1]![tx1]![bin]!;
      const top = v00 * (1 - fx) + v01 * fx;
      const bottom = v10 * (1 - fx) + v11 * fx;
      out[y * width + x] = top * (1 - fy) + bottom * fy;
    }
  }

  return { width, height, data: out };
}

function clamp01(v: number): number {
  return v < 0 ? 0 : v > 1 ? 1 : v;
}
function clampInt(v: number, lo: number, hi: number): number {
  return v < lo ? lo : v > hi ? hi : v;
}

function buildTileLut(
  data: Float32Array,
  imageWidth: number,
  x0: number,
  y0: number,
  tileWidth: number,
  tileHeight: number,
  imageBoundWidth: number,
  imageBoundHeight: number,
  bins: number,
  clipLimit: number,
): Float32Array {
  const x1 = Math.min(x0 + tileWidth, imageBoundWidth);
  const y1 = Math.min(y0 + tileHeight, imageBoundHeight);
  const histogram = new Float64Array(bins);
  let count = 0;
  for (let y = y0; y < y1; y++) {
    for (let x = x0; x < x1; x++) {
      const bin = clampInt(Math.round(data[y * imageWidth + x]! * (bins - 1)), 0, bins - 1);
      histogram[bin]! += 1;
      count += 1;
    }
  }
  if (count === 0) {
    const identity = new Float32Array(bins);
    for (let i = 0; i < bins; i++) identity[i] = i / (bins - 1);
    return identity;
  }

  // Recorte del histograma (contrast limiting) y redistribución uniforme
  // del exceso entre todos los bins, como el CLAHE estándar.
  const clip = Math.max(1, clipLimit * (count / bins));
  let excess = 0;
  for (let i = 0; i < bins; i++) {
    if (histogram[i]! > clip) {
      excess += histogram[i]! - clip;
      histogram[i] = clip;
    }
  }
  const redistribution = excess / bins;
  for (let i = 0; i < bins; i++) histogram[i]! += redistribution;

  const lut = new Float32Array(bins);
  let cumulative = 0;
  const total = count;
  for (let i = 0; i < bins; i++) {
    cumulative += histogram[i]!;
    lut[i] = cumulative / total;
  }
  return lut;
}

/**
 * Heurístico de referencia: conserva la banda de columnas con mayor
 * varianza de intensidad, con un margen. **No** es el detector de ROI final
 * (exigiría su propio modelo entrenado) — ver `training/preprocess.py`,
 * mismo algoritmo, mismo motivo documentado allí.
 */
export function detectSpineRoi(image: GrayscaleImage, minWidthFraction = 0.3, marginFraction = 0.15): SpineRoi {
  const { width, height, data } = image;
  const columnVariance = new Float64Array(width);
  for (let x = 0; x < width; x++) {
    let sum = 0;
    let sumSq = 0;
    for (let y = 0; y < height; y++) {
      const v = data[y * width + x]!;
      sum += v;
      sumSq += v * v;
    }
    const mean = sum / height;
    columnVariance[x] = sumSq / height - mean * mean;
  }

  let maxVariance = 0;
  for (let x = 0; x < width; x++) if (columnVariance[x]! > maxVariance) maxVariance = columnVariance[x]!;
  const threshold = maxVariance * 0.2;

  let firstActive = -1;
  let lastActive = -1;
  for (let x = 0; x < width; x++) {
    if (columnVariance[x]! >= threshold) {
      if (firstActive === -1) firstActive = x;
      lastActive = x;
    }
  }
  if (firstActive === -1) {
    return { x0: 0, y0: 0, x1: width, y1: height };
  }

  let x0 = firstActive;
  let x1 = lastActive + 1;
  const margin = Math.floor((x1 - x0) * marginFraction);
  x0 = Math.max(0, x0 - margin);
  x1 = Math.min(width, x1 + margin);

  const minWidth = Math.floor(width * minWidthFraction);
  if (x1 - x0 < minWidth) {
    const center = Math.floor((x0 + x1) / 2);
    x0 = Math.max(0, center - Math.floor(minWidth / 2));
    x1 = Math.min(width, x0 + minWidth);
  }

  return { x0, y0: 0, x1, y1: height };
}

export interface ResampleResult {
  image: GrayscaleImage;
  transform: AffineTransform2D;
}

/**
 * Recorta a `roi` y remuestrea a `targetWidth × targetHeight` por
 * interpolación bilineal, devolviendo la `AffineTransform2D` que mapea el
 * espacio original al remuestreado (SPEC.md §8 Etapa 2). Mismo contrato que
 * `training/preprocess.py::resample_with_affine`.
 */
export function resampleWithAffine(image: GrayscaleImage, roi: SpineRoi, targetWidth: number, targetHeight: number): ResampleResult {
  const croppedWidth = roi.x1 - roi.x0;
  const croppedHeight = roi.y1 - roi.y0;
  if (croppedWidth <= 0 || croppedHeight <= 0) {
    throw new Error(`ROI vacía: ${JSON.stringify(roi)} sobre una imagen de ${image.width}x${image.height}.`);
  }

  const scaleX = targetWidth / croppedWidth;
  const scaleY = targetHeight / croppedHeight;
  const out = new Float32Array(targetWidth * targetHeight);

  for (let ty = 0; ty < targetHeight; ty++) {
    // Coordenada correspondiente en el espacio recortado (centro de píxel).
    const srcYCrop = (ty + 0.5) / scaleY - 0.5;
    const srcY = clampInt(Math.round(roi.y0 + srcYCrop), roi.y0, roi.y1 - 1);
    for (let tx = 0; tx < targetWidth; tx++) {
      const srcXCrop = (tx + 0.5) / scaleX - 0.5;
      const srcX = clampInt(Math.round(roi.x0 + srcXCrop), roi.x0, roi.x1 - 1);
      out[ty * targetWidth + tx] = bilinearSample(image, roi.x0 + srcXCrop, roi.y0 + srcYCrop, srcX, srcY);
    }
  }

  const cropTransform: AffineTransform2D = { scaleX: 1, scaleY: 1, offsetX: -roi.x0, offsetY: -roi.y0 };
  const resizeTransform: AffineTransform2D = { scaleX, scaleY, offsetX: 0, offsetY: 0 };
  return { image: { width: targetWidth, height: targetHeight, data: out }, transform: composeAffine(cropTransform, resizeTransform) };
}

/** Muestreo bilineal con `(fallbackX, fallbackY)` como respaldo cuando las
 * coordenadas continuas caen fuera de los límites de la imagen. */
function bilinearSample(image: GrayscaleImage, x: number, y: number, fallbackX: number, fallbackY: number): number {
  const x0 = Math.floor(x);
  const y0 = Math.floor(y);
  if (x0 < 0 || y0 < 0 || x0 + 1 >= image.width || y0 + 1 >= image.height) {
    return image.data[clampInt(fallbackY, 0, image.height - 1) * image.width + clampInt(fallbackX, 0, image.width - 1)]!;
  }
  const fx = x - x0;
  const fy = y - y0;
  const v00 = image.data[y0 * image.width + x0]!;
  const v01 = image.data[y0 * image.width + x0 + 1]!;
  const v10 = image.data[(y0 + 1) * image.width + x0]!;
  const v11 = image.data[(y0 + 1) * image.width + x0 + 1]!;
  const top = v00 * (1 - fx) + v01 * fx;
  const bottom = v10 * (1 - fx) + v11 * fx;
  return top * (1 - fy) + bottom * fy;
}
